### PR TITLE
Fix Additive Production

### PR DESCRIPTION
**Pull request by hacatu**
When production modifiers were made additive, some default values were not changed from 1 to 0 and other mistakes were made.  This should fix that

Also fixed frame tooltips for frames with negative modifiers

### CODE REVIEW

**Comment by hacatu** [comment]
See https://github.com/GTNewHorizons/Binnie/pull/33, https://github.com/GTNewHorizons/MagicBees/pull/30, https://github.com/GTNewHorizons/GTplusplus/pull/691

**Review by kuba6000 — APPROVED** [review]

**Review by kuba6000 — CHANGES_REQUESTED** [review]
Currently the color on the tooltip applies as it would be multiplicative modifier.
I would suggest to change the coloring to match the additive aspects.
(+0 should be gold look screenshot below)

image

File: src/main/java/forestry/apiculture/BeeHousingModifier.java [hunk 1]
```diff
@@ -49,8 +49,7 @@ public float getLifespanModifier(IBeeGenome genome, IBeeGenome mate, final float
 
     @Override
     public float getProductionModifier(IBeeGenome genome, final float currentModifier) {
-        float modifierValue = 0.0f;// the "base" value of 1.0 should come from Bee.java or BeeKeepingMode.jave, that is,
-                                   // it should be set at the call site
+        float modifierValue = 1.0f;
```
Comment (kuba6000) [inline src/main/java/forestry/apiculture/BeeHousingModifier.java]: That was actually right, the new production modifier is supposed to start at 0, therefore the normal apiary without frames should have -0.9 (not 0.1) by default and for magic apiary it should be -0.1 (not 0.9). Please revert this commit.

**Review by kuba6000 — CHANGES_REQUESTED** [review]

File: src/main/java/forestry/apiculture/genetics/Bee.java [hunk 2]
```diff
@@ -514,8 +514,8 @@ public ItemStack[] produceStacks(IBeeHousing housing) {
 
         // Bee genetic speed * beehousing * beekeeping mode
         float speed = genome.getSpeed();
-        float prodModifier = beeHousingModifier.getProductionModifier(genome, 1f)
-                * beeModeModifier.getProductionModifier(genome, 1f);
+        float prodModifier = beeHousingModifier.getProductionModifier(genome, 1f);
+        prodModifier += beeModeModifier.getProductionModifier(genome, prodModifier);
```
[the remark this answers is not in the document]
Comment (hacatu) [reply src/main/java/forestry/apiculture/genetics/Bee.java]: Fixed in https://github.com/GTNewHorizons/ForestryMC/pull/55/commits/96afda431fe7519fb2cd0e31dbefbcdd0335b196.  I agree that it looks better that way, but `IBeeModifier.getProductionModifier` expects the current production modifier as its last argument, so we have to provide it correctly and do an updating assignment `+=`

File: src/main/java/forestry/apiculture/genetics/Bee.java [hunk 3]
```diff
@@ -551,6 +551,9 @@ public ItemStack[] produceStacks(IBeeHousing housing) {
     }
 
     public static float getFinalChance(float chance, float beeSpeed, float productionModifier, float t) {
+        if (productionModifier + t <= 0f) {
+            return 0;
```
[the remark this answers is not in the document]
Comment (hacatu) [reply src/main/java/forestry/apiculture/genetics/Bee.java]: Done in https://github.com/GTNewHorizons/ForestryMC/pull/55/commits/96afda431fe7519fb2cd0e31dbefbcdd0335b196

File: src/main/java/forestry/apiculture/items/ItemHiveFrame.java [hunk 4]
```diff
@@ -56,7 +56,7 @@ public HiveFrameBeeModifier(float geneticDecay) {
 
         @Override
         public float getProductionModifier(IBeeGenome genome, float currentModifier) {
-            return (currentModifier < 10f) ? 2f : 1f;
+            return (currentModifier < 10.0f) ? 2.0f : 0.0f;
```
[the remark this answers is not in the document]
Comment (hacatu) [reply src/main/java/forestry/apiculture/items/ItemHiveFrame.java:59]: These apply the modifiers for the base forestry frames (impregnated, proven, etc).  They're supposed to have a modifier of 2 and we didn't touch any other positive modifiers, so I don't think we should nerf this

File: src/main/java/forestry/apiculture/BeeHousingModifier.java (hunk 1 above)
Comment (hacatu) [reply src/main/java/forestry/apiculture/BeeHousingModifier.java]: Meant to just click confirm but pushing changes manually removed that option.

Fixed in https://github.com/GTNewHorizons/ForestryMC/pull/55/commits/96afda431fe7519fb2cd0e31dbefbcdd0335b196

File: src/main/java/forestry/apiculture/items/ItemHiveFrame.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kuba6000) [reply src/main/java/forestry/apiculture/items/ItemHiveFrame.java:59]: > These apply the modifiers for the base forestry frames (impregnated, proven, etc). They're supposed to have a modifier of 2 and we didn't touch any other positive modifiers, so I don't think we should nerf this

We are adjusting all frames

File: src/main/java/forestry/apiculture/EventHandlerApiculture.java [hunk 5]
```diff
@@ -34,19 +34,43 @@ public void addFrameTooltip(ItemTooltipEvent event) {
...
+                                StringUtil.localize("frame.tooltip.lifespan")
+                                        + getModifierFormatted(lifespan, false, true));
+                    }
+                    float production = modifier.getProductionModifier(null, 0.0F);
+                    if (production != 0.0F) {
+                        tooltip.add(
+                                StringUtil.localize("frame.tooltip.production")
+                                        + getModifierFormatted(production, true, false));
+                    }
+                    float flowering = modifier.getProductionModifier(null, 1.0F);
```
Comment (kuba6000) [inline src/main/java/forestry/apiculture/EventHandlerApiculture.java:61]: Wrong modifier
```suggestion
                    float flowering = modifier.getFloweringModifier(null, 1.0F);
```

File: src/main/java/forestry/apiculture/EventHandlerApiculture.java [hunk 6]
```diff
@@ -34,19 +34,43 @@ public void addFrameTooltip(ItemTooltipEvent event) {
...
+                                        + getModifierFormatted(flowering, false, false));
+                    }
                     float decay = modifier.getGeneticDecay(null, 1.0F);
-                    tooltip.add(StringUtil.localize("frame.tooltip.geneticDecay") + getColorFormatted(decay));
+                    if (decay != 1.0F) {
+                        // lower genetic decay is better
+                        tooltip.add(
+                                StringUtil.localize("frame.tooltip.geneticDecay")
+                                        + getModifierFormatted(decay, false, true));
+                    }
```
Comment (kuba6000) [inline src/main/java/forestry/apiculture/EventHandlerApiculture.java:73]: Why are you hiding modifiers that does not change anything? (x1, +0) I think it looks worse now that they are hidden. Please revert.

**Review by kuba6000 — CHANGES_REQUESTED** [review]

File: src/main/java/forestry/apiculture/EventHandlerApiculture.java (hunk 5 above)
Comment (hacatu) [reply src/main/java/forestry/apiculture/EventHandlerApiculture.java]: Fixed in https://github.com/GTNewHorizons/ForestryMC/pull/55/commits/a868679d8ec93f2201ba65fea73e1c2f303f34bd

**Review by Runakai1 — APPROVED** [review]

**Comment by Dream-Master** [comment]
Approved

**Pull request merged**